core/src/nativeMain/kotlin/com.saveourtool.processbuilder/ProcessBuilderInternal.kt

```diff
@@ -0,0 +1,72 @@
+/**
+ * Utilities to run a process and get its result.
+ */
+
+package com.saveourtool.processbuilder
+
+import com.saveourtool.processbuilder.exceptions.ProcessTimeoutException
+import com.saveourtool.processbuilder.utils.isCurrentOsWindows
+
+import okio.Path
+import platform.posix.system
+
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.async
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.joinAll
+import kotlinx.coroutines.newSingleThreadContext
+import kotlinx.coroutines.runBlocking
+
+@Suppress(
+    "MISSING_KDOC_TOP_LEVEL",
+    "MISSING_KDOC_CLASS_ELEMENTS",
+    "MISSING_KDOC_ON_FUNCTION"
+)
+actual class ProcessBuilderInternal actual constructor(
+    private val stdoutFile: Path,
+    private val stderrFile: Path,
+    private val useInternalRedirections: Boolean
+) {
+    actual fun prepareCmd(command: String) = if (useInternalRedirections) {
+        "($command) >$stdoutFile 2>$stderrFile"
+    } else {
+        command
+    }
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    actual fun exec(
+        cmd: String,
+        timeOutMillis: Long
+    ): Int {
+        var status = -1
+
+        runBlocking {
+            val timeOut = async(newSingleThreadContext("timeOut")) {
+                delay(timeOutMillis)
+                destroy(cmd)
+                throw ProcessTimeoutException(timeOutMillis, "Timeout is reached: $timeOutMillis")
+            }
+
+            val command = async {
+                status = system(cmd)
+                timeOut.cancel()
+            }
+            joinAll(timeOut, command)
+        }
+
+        if (status == -1) {
+            error("Couldn't execute $cmd, exit status: $status")
+        }
+        return status
+    }
+
+    private fun destroy(cmd: String) {
```

Comment:
if my memory serves me right https://github.com/saveourtool/save-cli/issues/338

also, in your free time, please check for all relevant issues to PB from save-cli and copy them into this repo

Reply:
Moved issues from `save-cli` to this repo. Original issues will be closed as not planned after `save-cli` will use `process-builder-multiplatform` as dependency